ludwig/benchmarking/summary_dataclasses.py

```diff
@@ -0,0 +1,104 @@
+import os
+from dataclasses import dataclass
+from typing import List
+
+from globals import REPORT_JSON
+
+from ludwig.globals import MODEL_HYPERPARAMETERS_FILE_NAME
+from ludwig.modules.metric_registry import get_metric_classes
+from ludwig.utils.data_utils import load_json
+
+
+@dataclass
+class ExperimentSummary:
+    """Summary of metrics from one experiment.
+
+    experiment_local_directory: path containing the artifacts for the experiment.
+    output_feature_type: LudwigModel output feature type.
+    output_feature_name: LudwigModel output feature name.
+    metric_to_values: dictionary that maps from metric name to their values.
+    metric_names: names of metrics for the output feature.
+    empty: True if unable to load metrics.
+    """
+
+    experiment_local_directory: str
+
+    def __post_init__(self):
```

Comment:
Since we're reading from `model_hyperparameters.json` and `report.json` and not from user-defined data, I made use of `__post_init__` to retrieve relevant JSON files and build the data summary from there. These files are probably not very readable, so I'm not sure how we can make use of these dataclasses without adding a lot of code in other files. WDYT?

Reply:
Let me clarify what I'm proposing. Instead of putting this code in __post_init__, we should put that code in a separate library, and users can invoke the library instead of constructing an `ExperimentSummary` directly.

```python
@dataclass
class ExperimentSummary:
    """Summary of metrics from one experiment."""

    # Path containing the artifacts for the experiment.
    experiment_local_directory: str

    # True if unable to load metrics.
    empty: bool

    # LudwigModel output feature type.
    output_feature_type: Optional[str] = None

    # LudwigModel output feature name.
    output_feature_name: Optional[str] = None

    # Dictionary that maps from metric name to their values.
    metric_to_values: Optional[Dict] = None

    # Names of metrics for the output feature.
    metric_names: Optional[Set[str]] = None


def build_experiment_summary(experiment_local_directory) -> ExperimentSummary:
    """Returns an ExperimentSummary with empty=True if the directory cannot be read."""
    try:
        config = load_json(os.path.join(experiment_local_directory, MODEL_HYPERPARAMETERS_FILE_NAME))
        report = load_json(os.path.join(experiment_local_directory, REPORT_JSON))
    except Exception:
        return ExperimentSummary(experiment_local_directory=experiment_local_directory, empty=True)

    performance_metrics = report["evaluate"]["performance_metrics"]
    output_feature_type: str = self.config["output_features"][0]["type"]
    output_feature_name: str = self.config["output_features"][0]["name"]
    metric_dict = performance_metrics[self.output_feature_name]
    full_metric_names = get_metric_classes(self.output_feature_type)
    metric_to_values: dict = {
        metric_name: metric_dict[metric_name] for metric_name in full_metric_names if metric_name in metric_dict
    }
    metric_names: set = set(self.metric_to_values.keys())

    return ExperimentSummary(
        experiment_local_directory=experiment_local_directory, 
        output_feature_type=output_feature_type,
        output_feature_name=output_feature_name,
        metric_dict=metric_dict,
        full_metric_names=full_metric_names,
        metric_to_values=metric_to_values,
        metric_names=metric_names
    )
```

This allows all of the fields of the dataclass to be cleanly declared, while still enabling a clean way for users to instantiate an `ExperimentSummary`.